server/blob/blob-api/src/main/java/org/apache/james/blob/api/Store.java

```diff
@@ -0,0 +1,115 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.api;
+
+import java.io.InputStream;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.james.util.FluentFutureStream;
+
+import com.google.common.collect.ImmutableMap;
+
+public interface Store<T> {
+    class BlobType {
+        private final String name;
+
+        public BlobType(String name) {
+            this.name = name;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        @Override
+        public final boolean equals(Object o) {
+            if (o instanceof BlobType) {
+                BlobType blobType = (BlobType) o;
+
+                return Objects.equals(this.name, blobType.name);
+            }
+            return false;
+        }
+
+        @Override
+        public final int hashCode() {
+            return Objects.hash(name);
+        }
+    }
+
+    interface Encoder<T> {
+        Map<BlobType, InputStream> encode(T t);
+    }
+
+    interface Decoder<T> {
+        void validateInput(Collection<BlobType> input);
+
+        T decode(Map<BlobType, byte[]> streams);
+    }
+
+    CompletableFuture<Map<BlobType, BlobId>> save(T t);
```

Comment:
Store<T> is supposed to be used to save a T and read a T from the BlobStore.
BlobStore return a BlobId for each Blob it stores.
Encoder can change the way T is serialized (it can compress things, it can encrypt, it can split, etc) so it may not really return a BlobId.
We can either assume we have one or more BlobId for a save or introduce a different object to allow for more flexibility.
I don't see any usage of BlobType outside of the implementation because the caller should not even know there are several BlobId.

Reply:
should we introduce a new BlobId (ComposedBlobId) which could be returned by this API?
From the rest of the world we still handle only blobIds, and here we are able to get several objects at the same time

Reply:
I mean the current Map<BlobType, BlobId> could be encapsulated into a BlobId implementation.